include/exadg/incompressible_navier_stokes/spatial_discretization/turbulence_model.cpp

```diff
@@ -545,11 +624,152 @@ TurbulenceModel<dim, Number>::sigma_model(scalar const & filter_width,
+  viscosity += dealii::make_vectorized_array<Number>(viscosity_newtonian_limit);
+}
+
+template<int dim, typename Number>
+void
+ViscosityModel<dim, Number>::cross_generalized_newtonian_model(scalar const & shear_rate,
+                                                               scalar &       viscosity) const
+{
+  // eta = eta_oo + (eta_0 - eta_oo) * [k + (l * y)^a]^[(n-1)/a]
+  // with k = 1 and n = 1 - a
+  // eta = eta_oo + (eta_0 - eta_oo) * [1 + (l * y)^a]^(-1)
+  viscosity =
+    shear_rate * dealii::make_vectorized_array<Number>(generalized_newtonian_model_data.lambda);
+
+  viscosity =
+    std::pow(viscosity, dealii::make_vectorized_array<Number>(generalized_newtonian_model_data.a));
+
+  viscosity += dealii::make_vectorized_array<Number>(1.0);
+  viscosity = dealii::make_vectorized_array<Number>(1.0) / viscosity;
+
+  viscosity *= dealii::make_vectorized_array<Number>(
+    generalized_newtonian_model_data.viscosity_upper_limit - viscosity_newtonian_limit);
+  viscosity += dealii::make_vectorized_array<Number>(viscosity_newtonian_limit);
+}
+
+template<int dim, typename Number>
+void
+ViscosityModel<dim, Number>::simplified_cross_generalized_newtonian_model(scalar const & shear_rate,
+                                                                          scalar & viscosity) const
+{
+  // eta = eta_oo + (eta_0 - eta_oo) * [k + (l * y)^a]^[(n-1)/a]
+  // with k = 1, a = 1 and n = 0
+  // eta = eta_oo + (eta_0 - eta_oo) * [1 + l * y]^(-1)
+  viscosity =
+    shear_rate * dealii::make_vectorized_array<Number>(generalized_newtonian_model_data.lambda);
+
+  viscosity += dealii::make_vectorized_array<Number>(1.0);
+
+  viscosity = dealii::make_vectorized_array<Number>(1.0) / viscosity;
+
+  viscosity *= dealii::make_vectorized_array<Number>(
+    generalized_newtonian_model_data.viscosity_upper_limit - viscosity_newtonian_limit);
+  viscosity += dealii::make_vectorized_array<Number>(viscosity_newtonian_limit);
+}
+
+template<int dim, typename Number>
+void
+ViscosityModel<dim, Number>::power_law_generalized_newtonian_model(scalar const & shear_rate,
+                                                                   scalar &       viscosity) const
+{
+  // eta = eta_oo + (eta_0 - eta_oo) * [k + (l * y)^a]^[(n-1)/a]
+  // with k = 0
+  // eta = eta_oo + (eta_0 - eta_oo) * (l * y)^(n-1)
+  viscosity =
+    shear_rate * dealii::make_vectorized_array<Number>(generalized_newtonian_model_data.lambda);
+
+  viscosity =
+    std::pow(viscosity,
+             dealii::make_vectorized_array<Number>(generalized_newtonian_model_data.n - 1.0));
+
+  viscosity *= dealii::make_vectorized_array<Number>(
+    generalized_newtonian_model_data.viscosity_upper_limit - viscosity_newtonian_limit);
+  viscosity += dealii::make_vectorized_array<Number>(viscosity_newtonian_limit);
+}
 
-template class TurbulenceModel<3, float>;
-template class TurbulenceModel<3, double>;
+template class ViscosityModel<2, float>;
```

Comment:
I realized just now, that some of the turbulence models are only defined in 2D, but those are asserted anyways.